core/shared/src/main/scala/cats/effect/IO.scala

```diff
@@ -588,9 +589,19 @@ sealed abstract class IO[+A] private () extends IOPlatform[A] {
   def onCancel(fin: IO[Unit]): IO[A] =
     IO.OnCancel(this, fin)
 
+  @deprecated("Use onError with PartialFunction argument", "3.6.0")
   def onError(f: Throwable => IO[Unit]): IO[A] =
     handleErrorWith(t => f(t).voidError *> IO.raiseError(t))
```

Comment:
Can we delegate this to the other method?
```suggestion
    handleErrorWith(f: PartialFunction[Throwable, IO[Unit]])
```

Reply:
mm, I'm not sure I understand, I think your suggestion shouldn't be compiled?
Or do you mean use the new `onError`?
btw, I'm about to go to bed, will think more and answer tomorrow.

Reply:
> I think your suggestion shouldn't be compiled? Or do you mean use the new `onError`?

Oops, I made a typo. But I think you are right, this won't compile as-is 🤔 

Basically we should avoid to duplicate the `onError` implementation. We can share the implementation by delegating one `onError` to the other `onError`, by converting `Function` to `PartialFunction` or vice versa.

Reply:
Fixed this in [f4f505c](https://github.com/typelevel/cats-effect/pull/4121/commits/f4f505c8e70adb930662d33a22e0415ff6018e07)

Another way would be 
```diff
-    handleErrorWith(t => pf.applyOrElse(t, (_: Throwable) => IO.unit) *> IO.raiseError(t))
+    onError(t => pf.applyOrElse(t, (_: Throwable) => IO.unit))
```

but it feels awkward as We used a function that We just deprecated it.